spirv_msl.cpp

```diff
@@ -10672,6 +11659,55 @@ bool CompilerMSL::OpCodePreprocessor::handle(Op opcode, const uint32_t *args, ui
 		uint32_t result_type = args[0];
 		uint32_t id = args[1];
 		uint32_t ptr = args[2];
+
+		// Fix tessellation patch function processing
+		if (compiler.get_execution_model() == ExecutionModelTessellationControl)
+		{
+			uint32_t source_id = args[3];
+			bool is_indexed_by_invocation =
+			    variables_indexed_by_invocation.find(ptr) != variables_indexed_by_invocation.end() ||
+			    invocation_ids.find(source_id) != invocation_ids.end();
+			if (!is_indexed_by_invocation)
+			{
+				auto *var = compiler.maybe_get_backing_variable(source_id);
+				if (var)
+					source_id = var->self;
+
+				// Only interested in standalone builtin variables.
+				if (compiler.has_decoration(source_id, DecorationBuiltIn))
+				{
+					auto builtin = static_cast<BuiltIn>(compiler.get_decoration(source_id, DecorationBuiltIn));
+					switch (builtin)
+					{
+					case BuiltInInvocationId:
+						is_indexed_by_invocation = true;
+						break;
+					default:
+						break;
+					}
+				}
+			}
+
+			if (is_indexed_by_invocation)
+			{
+				if (passed_control_barrier)
+				{
+					auto *var = compiler.maybe_get_backing_variable(ptr);
```

Comment:
Ok, so I have a feeling this is an unnecessary hack for invalid SPIR-V. Looking at `shaders-msl/asm/tesc/tess-factor-must-be-threadgroup.asm.tesc`.

GLSL output is:
```
...
    FFlatTessellationHSToDS temp_var_hullMainRetVal[3]; // <-- Huh?
    temp_var_hullMainRetVal[gl_InvocationID] = FFlatTessellationHSToDS(param_var_I[gl_InvocationID], _125, Material.Material_ScalarExpressions[0].x, 1.0);
    barrier();
    if (gl_InvocationID == 0u)
    {
        vec4 _151 = _85;
        _151.x = 0.5 * (temp_var_hullMainRetVal[1u].TessellationMultiplier + temp_var_hullMainRetVal[2u].TessellationMultiplier);
        vec4 _157 = _151;
        _157.y = 0.5 * (temp_var_hullMainRetVal[2u].TessellationMultiplier + temp_var_hullMainRetVal[0u].TessellationMultiplier);
        vec4 _162 = _157;
        _162.z = 0.5 * (temp_var_hullMainRetVal[0u].TessellationMultiplier + temp_var_hullMainRetVal[1u].TessellationMultiplier);
        vec4 _169 = _162;
        _169.w = 0.333000004291534423828125 * ((temp_var_hullMainRetVal[0u].TessellationMultiplier + temp_var_hullMainRetVal[1u].TessellationMultiplier) + temp_var_hullMainRetVal[2u].TessellationMultiplier);
...
```
This is the kind of thing the MSL patch will check for, and that array is forced into a groupshared array instead, but the SPIR-V is just wrong. For the other cases where patch constant phase reads from Output variables, they are already in device memory, not threadgroup, and this "fix" actually breaks those shaders.

This is very likely a DXC/glslang bug if anything.

Reply:
Consider a reduced case for GLSL which does something similar:

```
#version 450

layout(vertices = 2) out;

layout(location = 0) patch out vec4 vFoo;
layout(location = 1) out vec4 vOutput[2];

void main()
{
	vOutput[gl_InvocationID] = vec4(1.0);
	barrier();
	if (gl_InvocationID == 0)
	{
		vFoo = vOutput[0] + vOutput[1];
		gl_TessLevelInner[0] = 1.0;
		gl_TessLevelOuter[1] = 2.0;
	}
}
```

This is now converted into:
```
#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct main0_patchOut
{
    float4 vFoo;
};

kernel void main0(uint gl_InvocationID [[thread_index_in_threadgroup]], uint gl_PrimitiveID [[threadgroup_position_in_grid]], constant uint* spvIndirectParams [[buffer(29)]], device main0_patchOut* spvPatchOut [[buffer(27)]], device MTLQuadTessellationFactorsHalf* spvTessLevel [[buffer(26)]])
{
    threadgroup float4 vOutput[2]; // <-- Wrong.
    device main0_patchOut& patchOut = spvPatchOut[gl_PrimitiveID];
    vOutput[gl_InvocationID] = float4(1.0);
    threadgroup_barrier(mem_flags::mem_device | mem_flags::mem_threadgroup);
    if (gl_InvocationID == 0)
    {
        patchOut.vFoo = vOutput[0] + vOutput[1];
        spvTessLevel[gl_PrimitiveID].insideTessellationFactor[0] = half(1.0);
        spvTessLevel[gl_PrimitiveID].edgeTessellationFactor[1] = half(2.0);
    }
}
```